### DISCUSSION TITLE
Escape sequences in raw string literals

### DISCUSSION THREAD

**Discussion by phizch** [dotnet/csharplang#6178] [Language Ideas] [answered]
Should it be possible to have escape sequences in [raw string literals](https://github.com/dotnet/csharplang/blob/main/proposals/raw-string-literal.mdl)?
It wouldn't be 100% _raw_ anymore, but it's already got interpolation..
```cs
string doesNotWork = """Test\x0020Test""";     // Test\x0020Test
```

## Why?

This would be useful for including Unicode characters that doesn't have a glyph in the editor's font. It could also be useful for characters that are visually very similar to each other.
```cs
string similarCharacters = "\x003a \x02f8 \xa789  \x2236";
// ": ꞉ ꞉ ∶ " -- Colon, Modifier Letter Raised Colon, Modifier Letter Colon, Ratio
```

### Suggested syntax:
Treat _n_ number of slashes as a escape sequence, where _n_ is the number of `"`  enclosing the string. 
E.g. `"""\\\x0020"""`.
```cs
string proposed = """Test\\\x0020Test""";    // Test Test
```

## Alternatives
Currently it's possible to include escape sequences in raw string literals by using string interpolation:
```cs
string correct = $"""Test{"\x0020"}Test""";  // Test Test
```
To be honest, this works fine, but it might be good to communicate this option to users of RSLs. 
With the new-ish constant string interpolation, the compiler should produce an identical byte sequence in the executable anyway, so it wouldn't affect performance.

**Comment by CyrusNajmabadi** [comment, marked as answer]
> Should it be possible to have escape sequences in [raw string literals](https://github.com/dotnet/csharplang/blob/main/proposals/raw-string-literal.mdl)?

No.  It defeats the entire purpose and motivation of raw-strings.  WHich is *no escaping*.  The moment you may have escape sequences, then you need a way to escape out of that.

**Reply by phizch** [reply]
Verbatim strings doesn't have escaping, but that's (imo) more for the convenience of writing e.g. paths, without having to escape all the backslashes, and uses `""` as an "escape" for `"`. But since raw strings have another mechanism to avoid having to escape commonly used characters.
I'm not sure I understand what you mean with "you need a way to escape out of that.".
If you mean escaping `\`, then I envision it the same way as with `{` with interpolated RSLs, except I was using the number of `"` instead of the number of `$` for interpolated.
Eg.
```cs
"""Test\x02f8 A test."""     // Test\x02f8 A test.
"""Test\\\0x02f8 A test."""  // Test꞉ A test.
"""Test\\\\0x02f8 A test."""  // Test\꞉ A test.
string x = "\x02f8";
$$"""Test{{{x}}} A test.""" // Test{:} A test.
$$"""Test{{x}} A test.""" // Test: A test.
```

**Comment by CyrusNajmabadi** [comment]
> Treat n number of slashes as a escape sequence, where n is the number of " enclosing the string.
E.g. `"""\\\x0020"""`.

This would break existing code where that has different meaning already.

**Reply by phizch** [reply]
Out of curiosity, where? BTW, seems like one of the slashes disappeared in your quote. Bug in GH? :-)

**Reply by CyrusNajmabadi** [reply]
Anywhere :)

The code you wrote is legal today, and has a particular meaning.  We cannot cahnge that meaning or it will break *any* code using that pattern.

That said, i want to be clear: this is antithetical to raw-strings :)  The point is to not need to escape.  If you want to escape things... use a normal string.  They handle escapes just fine.

**Reply by jnm2** [reply]
> This would break existing code where that has different meaning already.

The feature has only been available for a month and under a preview flag, and breaking existing code (defined this way) has been done before. Most recently, when rolling back the addition of `!!` to the language.

**Comment by phizch** [comment]
I've updated the alternative. The original `$"""{'\x20'}"""` wouldn't work for const strings, since chars can't be compile time interpolated.

**Reply by CyrusNajmabadi** [reply]
you can do: `$"""{"\x20"}"""`

**Reply by phizch** [reply]
The updated code in my original post was from from `'\x0020`  to `"\x0020"`.

Yeah, I used \x0020 instead of \x20 since it's more robust for general use. Eg. `"\x20a test` would be treated as `"\x020a test"`. Also, I used \x20 since it's simply `' '`. Escaping it obviously doesn't make much sense :-)

**Comment by jnm2** [comment]
@phizch I've had the same thought as you via organic usage of this feature, but I actually prefer the interpolation alternative over the proposed repeated backslash (cool as that idea was). Unless the proposed syntax is enough orders of magnitude better than the alternative, it seems worse to complicate the principle of "no escaping" and make these strings harder to think about whenever a backslash appears. Interpolation itself was a complication that also brought a huge step forward in enabling raw string literals to be used in common real-world situations; this wouldn't be the case again with backslashes.

**Reply by jnm2** [reply]
Another thought is that `\ud83e\ude84` is not really as nice to look at within a string as `{MagicWandEmoji}` anyway.

(Okay, bad example because here you could just put the emoji inline, but the point stands for unicode characters that you don't want to put in the string directly.)
